isisd/isis_ldp_sync.c

```diff
@@ -0,0 +1,781 @@
+ *
+ * GNU Zebra is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation; either version 2, or (at your option) any
+ * later version.
+ *
+ * GNU Zebra is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; see the file COPYING; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include <zebra.h>
+#include <string.h>
+
+#include "monotime.h"
+#include "memory.h"
+#include "thread.h"
+#include "prefix.h"
+#include "table.h"
+#include "vty.h"
+#include "command.h"
+#include "plist.h"
+#include "log.h"
+#include "zclient.h"
+#include <lib/json.h>
+#include "defaults.h"
+#include "ldp_sync.h"
+
+#include "isisd/isis_constants.h"
+#include "isisd/isis_common.h"
+#include "isisd/isis_flags.h"
+#include "isisd/isis_circuit.h"
+#include "isisd/isis_lsp.h"
+#include "isisd/isis_pdu.h"
+#include "isisd/isis_network.h"
+#include "isisd/isis_misc.h"
+#include "isisd/isis_constants.h"
+#include "isisd/isis_adjacency.h"
+#include "isisd/isis_dr.h"
+#include "isisd/isisd.h"
+#include "isisd/isis_csm.h"
+#include "isisd/isis_events.h"
+#include "isisd/isis_te.h"
+#include "isisd/isis_mt.h"
+#include "isisd/isis_errors.h"
+#include "isisd/isis_tx_queue.h"
+#include "isisd/isis_nb.h"
+#include "isisd/isis_ldp_sync.h"
+
+extern struct zclient *zclient;
+
+/*
+ * LDP-SYNC msg between IGP and LDP
+ */
+int isis_ldp_sync_state_update(struct ldp_igp_sync_if_state state)
+{
+	struct interface *ifp;
+	struct isis_circuit *circuit = NULL;
+	struct isis_area *area;
+	struct listnode *node;
+	struct isis *isis = isis_lookup_by_vrfid(VRF_DEFAULT);
+
+	/* if isis is not enabled or LDP-SYNC is not configured ignore */
+	if (!isis ||
+	    !CHECK_FLAG(isis->ldp_sync_cmd.flags, LDP_SYNC_FLAG_ENABLE))
+		return 0;
+
+	/* lookup circuit */
+	ifp = if_lookup_by_index(state.ifindex, VRF_DEFAULT);
+	if (ifp == NULL)
+		return 0;
+
+	for (ALL_LIST_ELEMENTS_RO(isis->area_list, node, area)) {
+		circuit = circuit_lookup_by_ifp(ifp, area->circuit_list);
+		if (circuit != NULL)
+			break;
+	}
+
+	/* if isis is not enabled or LDP-SYNC is not configured ignore */
+	if (circuit == NULL ||
+	    !CHECK_FLAG(isis->ldp_sync_cmd.flags, LDP_SYNC_FLAG_ENABLE))
+		return 0;
+
+	/* received ldp-sync interface state from LDP */
+	ils_debug("ldp_sync: rcvd %s from LDP if %s",
+		  state.sync_start ? "sync-start" : "sync-complete", ifp->name);
+	if (state.sync_start)
+		isis_ldp_sync_if_start(circuit, false);
+	else
+		isis_ldp_sync_if_complete(circuit);
+
+	return 0;
+}
+
+int isis_ldp_sync_announce_update(struct ldp_igp_sync_announce announce)
```

Comment:
The name is based on LDP announcing that it is up to ISIS.   If LDP crashes and starts up again it will send this announce message then also.   The announce message is used by ISIS to know that LDP is starting up so it must set the metric of all interfaces running ldp-sync to LSInfinity.   ISIS will then wait for LDP to provide an indication that it has exchanged all LDP labels between peers before it can restore cost on the interface.

Reply:
fixed